### ISSUE TITLE
Oppo A77 - Some texts gets cut-off

### ISSUE THREAD

**Issue by mykelaballe** [facebook/react-native#15114]
### Environment

1. `react-native -v`: 0.34 through 0.59
2. `node -v`:4.4.7
3. `npm -v`: 4.5.0
4. `yarn --version`:

- Target Platform: Android
- Development Operating System: Windows
- Build tools: The issue is specific to Oppo A77 model. Note that upgrading to latest react-native version is not an option now due to breaking changes and time constraints

### Steps to Reproduce

There's no necessary steps to produce these bug. You just have to run the app either in dev or production

### Expected Behavior

Texts should not be cut-off

### Actual Behavior

Some texts are cut-off as shown below. These cut-off issue doesnt occur in iphone and other phone models except Oppo A77

Screenshots:
![image](https://user-images.githubusercontent.com/22584900/28396742-ed4cf36a-6d2f-11e7-9fc0-d51a2885f0ca.png)

![image](https://user-images.githubusercontent.com/22584900/28396785-2f968696-6d30-11e7-8681-d992874ce9ed.png)

Those screenshots are just few of the texts that are cut-off

Sorry i can't provide a link to the project

**Comment by hramos** [comment]
Thanks for opening the issue! It does not appear like a community member will be able to reliably reproduce this issue. This may be for several reasons; perhaps it affects a particular app but a minimal repro has not been provided, or the issue may be sporadic. As it happens, we need a concrete set of steps that can demonstrably reproduce the issue as this will allow your fellow community members to validate a fix. We'll close the issue for now, but feel free to submit a new issue once you're able to reliably reproduce the issue locally. Thanks for your understanding!

**Comment by johnwook** [comment]
@mykelaballe cc @hramos We've had similar issue(some texts getting cut off in Oppo device). Although it's not a complete solution, we found some hacky-way to solve this problem.

in short, (at least in our case) just adding empty `space` at the end of the text seems work. my colleague @saystone will explain in detail :)

for example, use `1.5.3 ` instead of `1.5.3`(space at the end of the string)

**Comment by zouliang073** [comment]
Need to set a fixed width for `Text`

**Comment by asdacap** [comment]
As a side note, I also have this issue. It seems to appear on multiple Oppo devices. Probably something to do with fonts.

**Comment by jemmyphan** [comment]
@johnwook 
`in short, (at least in our case) just adding empty space at the end of the text seems work. my colleague @saystone will explain in detail :)`

do you still use this workaround?

**Comment by johnwook** [comment]
@jemmyphan Hello! Unfortunately yes :( . As our OPPO user grows, our team think we should SOLVE this problem right. For now there is no progress. If some arises, I'll let you know by commenting here. If you find some better solution please let me know :)

**Comment by jemmyphan** [comment]
@johnwook yeah, oppo user's grow so fast, anyway didn't expect it to be replied this fast. thanks and appreciate it a lot. ;)

**Comment by yeswanth** [comment]
I have faced similar issues in OnePlus 5. These issues only happen in a OnePlus 5 phone when you have chosen "OnePlus Slate" font in Settings. If you change the font to Roboto, the issues don't occur. I'm gonna ship Roboto font along with my app in the next version to go around this.

**Comment by asdacap** [comment]
@yeswanth Can you share how to ship the Roboto font along the app? Will all `Text` component automatically use the font or we'll have to make a wrapper `Text` component?

**Comment by yeswanth** [comment]
Hi @asdacap You can download [roboto font](https://fonts.google.com/?selection.family=Roboto) and add it to your project(As shown in this [link](https://medium.com/react-native-training/react-native-custom-fonts-ccc9aacf9e5e)). Unfortunately, you will have to create a wrapper for all `Text` in order for this to work.

**Comment by hkxicor** [comment]
this solution works for me, you can try this 
`
import React from 'react';
import { 
    StyleSheet,
    Text,
} from 'react-native';

const styles = StyleSheet.create({
    defaultFontFamily: {
        fontFamily: 'lucida grande',
    }
});

const oldRender = Text.prototype.render;
Text.prototype.render = function (...args) {
    const origin = oldRender.call(this, ...args);
    return React.cloneElement(origin, {
        style: [origin.props.style, styles.defaultFontFamily]
    });
};
`

**Comment by gedeagas** [comment]
@hkxicor trying your suggestion. will edit after the result

**Comment by asdacap** [comment]
I can say that it worked in my case.

**Comment by jslok** [comment]
@hkxicor do I need to place this code on every one of my js files with affected text or will it solve the issue for my entire app if I only put it in my main app.js?
```
const oldRender = Text.prototype.render;
Text.prototype.render = function (...args) {
const origin = oldRender.call(this, ...args);
return React.cloneElement(origin, {
style: [origin.props.style, styles.defaultFontFamily]
});
};
```

I'd test it myself but I do not personally have an oppo phone to reproduce on. Thank you

**Comment by hkxicor** [comment]
@jslok just put it on app.js it will work.

**Comment by jslok** [comment]
I tried the code but it appears to break react-native-vector-icons. All icons get replaced by X.

**Comment by hkxicor** [comment]
@jslok please update hare if you find any solution

**Comment by justjavac** [comment]
@hkxicor

```diff
 import React from 'react';
 
 import { StyleSheet, Text } from 'react-native';
 
 const styles = StyleSheet.create({
   defaultFontFamily: {
     fontFamily: 'lucida grande',
   },
 });
 
 export default function fixOppoTextCutOff() {
   const oldRender = Text.prototype.render;
   Text.prototype.render = function render(...args) {
     const origin = oldRender.call(this, ...args);
     return React.cloneElement(origin, {
-      style: [origin.props.style, styles.defaultFontFamily],
+      style: [styles.defaultFontFamily, origin.props.style],
     });
   };
 }
```

**Comment by jslok** [comment]
Can anybody else confirm if justjavac's solution above still fixes the issue of text getting cutoff? I don't have an oppo device to test, but I can say it does not cause react-native-vector-icons to break.

Also I have found OnePlus (not sure if some or all) devices also have this same issue.

**Comment by jslok** [comment]
Confirmed working

**Comment by justjavac** [comment]
@jslok 

Perhaps because OnePlus does not have the Lucida Grande font installed by default, you can manually install the Lucida Grande font in your react-native project.

My solution above is to solve "All icons get replaced by X". If there is still a text cutoff problem, check your fonts.

**Comment by jslok** [comment]
@justjavac
I meant the original issue of text getting cutoff exists on Oneplus devices just like Oppo devices. Your edit solves it on both, and also does not break react-native-vector-icons. Thank you and also thanks to @hkxicor

**Comment by joelandplus** [comment]
Setting the following seems to fix the issue:

<Text style={{fontFamily: ""}}>

**Comment by mohammadavood** [comment]
I have similar issue in LG G3 and react-native v0.53.0 
it looks like is a bug in react native that react support team doesn't like accept and fix it.
i solved it temporary by deleting  ((alignItems: 'center')) from the component style

**Comment by ou2s** [comment]
This issue should be open... Anyway, thanks for the workaround guys.

**Comment by mykelaballe** [comment]
thank God people took notice of this issue already. quite sometime. thanks guys

**Comment by parth3724** [comment]
@hkxicor your solution is working... but I don't understand the fix.. Fonts in my app are still Roboto and not lucida grande. I have not set fontfamily to <Text> ...

**Comment by WMisiedjan** [comment]
Subscribed, we're facing similar issues wile developing multiple apps!

**Comment by ramosalx1504** [comment]
For me the best solution is to specify the "fontFamily" for the global <Text> component in my application, so, no matter what typography the user has on their device, the application will be displayed as I designed it.

** If not specified, it seems that the <Text> component uses the device's default, and here come the problems (One plus 5T slate default font for example).

**Comment by wenkangzhou** [comment]
If some devices not install 'lucida grande',it can be work or crash?

**Comment by justjavac** [comment]
@wenkangzhou Not crash, it will fall back to using the default font

**Comment by wenkangzhou** [comment]
@justjavac ,good,thx.

**Comment by rpavlovs** [comment]
Same issue on iPhone X with `0.55.4`, but it seems to show up only after I browse in the app for a bit.

**Comment by hramos** [comment]
@rpavlos does it repro on 0.56?

**Comment by craigdanj** [comment]
We had the same problem with a one plus 3. Changing the font to Roboto fixed this.

as @yeswanth had suggested.

**Comment by byCedric** [comment]
I can confirm that this happens on React Native `0.55.x` with various custom fonts on Android. I haven't tried using the latest `0.56.x`, will try it when I can use it in Expo! For us the problem caused the React Navigation's header title being cut off with an ellipsis. If you have the same problem, you can simply set `width: '100%'` to this header title as a temporary workaround.

**Comment by Fitzpasd** [comment]
We can repro this on Android with RN `0.55.4`, **without using a custom font**.  None of the suggested 'fixes'/workarounds in this thread work

**Comment by hramos** [comment]
Please let us know if this can be reproduced in 0.56. If you've only tested on earlier releases, no need to +1.

**Comment by Fitzpasd** [comment]
Yes, it repos on 0.56.

**Comment by Eyosias5** [comment]
Its been a year and this issue is still here on the Oneplus 6 too.

**Comment by craigdanj** [comment]
So this goes away if you remove the one plus font called slate. Switch to
robot and you should see this go away.

**Comment by Eyosias5** [comment]
Yeah that worked. But wouldn't that mean I have to tell everyone who has
got a OnePlus to turn off the Slate font when they use my app? Isn't there
a way to force the device to change the font as you load the app?

**Comment by craigdanj** [comment]
I'm not an expert but can't you just set the font for your app to some
other font like roboto? That would override the system font.

**Comment by Eyosias5** [comment]
Yep that works great. Thanks :)

**Comment by hramos** [comment]
So, what can we do to fix this in the framework itself? Is React Native making a false assumption about the system font, and can we address this via a PR? Or is this something any developer might run into when targeting these devices even without RN?

**Comment by justjavac** [comment]
In an APP that is not developed using RN, this problem does not exist. I wrote a simple demo using Flutter and there is no such problem. Maybe this is a RN's problem.

**Comment by ou2s** [comment]
My 2 cents even if I don't have an affected model in my hands:
- Since it has to do with layout measurement, maybe we should have a look at the layout engine behind RN, [Yoga](https://github.com/facebook/yoga) and open an issue there.
- I suspect that Yoga is doing all its measurement with the default system font, but it seems the customized version of Android on these models replaces the system font with another one, making the layout calculation for the label size wrong.

**Comment by rpavlovs** [comment]
The cut-off line seems to slowly move further throughout the app use.
See the recording: https://streamable.com/59hx4

**Comment by vitorreis** [comment]
it seems to be related to #21729

**Comment by gagangoku** [comment]
Latest React Native does not have Text.prototype.render, use Text.render instead !

See https://stackoverflow.com/questions/35255645/how-to-set-default-font-family-in-react-native/47350760#47350760

**Comment by gagangoku** [comment]
Question: Is lucida grande font available on all android devices ? What happens if it's not ?

**Comment by SamMatthewsIsACommonName** [comment]
Hi @gauravds thanks for the solution. I'm slightly confused by the use. Is this something we import as a replacement component for <Text> element or do we simply import this file once in our app to cause a global change?
Thanks

**Comment by gauravds** [comment]
@SamMatthewsIsACommonName Simply import this file once in your app to cause a global change to all Text elements. One shot for all.  😄

**Comment by SamMatthewsIsACommonName** [comment]
Awesome thanks!

**Comment by LuisUrrutia** [comment]
Same issue with `0.59.0`

**Comment by rammmukul** [comment]
For people using expo as we do, this worked for us

```
import { setCustomTextInput, setCustomText } from "react-native-global-props";

import { Font } from "expo";

export default function () {
  Font.loadAsync({
    HelveticaNeue: require("../Assets/fonts/HelveticaNeue.ttf"),
    "HelveticaNeue Bold": require("../Assets/fonts/HelveticaNeue-Bold.ttf")
  }).then(() => {
    const customTextInputProps = {
      underlineColorAndroid: "rgba(0,0,0,0)"
    };

    const customTextProps = {
      style: {
        fontSize: 14,
        fontFamily: "HelveticaNeue"
      }
    };
    setCustomText(customTextProps);
    setCustomTextInput(customTextInputProps);
  });
} 
```

**Comment by nbstr** [comment]
> this solution works for me, you can try this
> 
> `import React from 'react';`
> `import { `
> ` StyleSheet,` Text,` `} from 'react-native';`
> 
> `const styles = StyleSheet.create({`
> ` defaultFontFamily: {`
> ` fontFamily: 'lucida grande',`
> ` }`
> `});`
> 
> `const oldRender = Text.prototype.render;`
> `Text.prototype.render = function (...args) {`
> ` const origin = oldRender.call(this, ...args);`
> ` return React.cloneElement(origin, {`
> ` style: [origin.props.style, styles.defaultFontFamily]`
> ` });`
> `};`

Unhandled JS Exception: Cannot read property 'render' of undefined

**Comment by snios** [comment]
No news on when this will be fixed? Almost 2 years old now....

**Comment by hramos** [comment]
@snios we'd be happy to take a look at a pull request that fixes this. Let me know if you're interested.

**Comment by snios** [comment]
@hramos thought there would be att least some one working on the development of this framework from Facebooks side. 

And sorry but i just started using it so i probably wont figure it out in a few years.

**Comment by gaodeng** [comment]
> And sorry but i just started using it so i probably wont figure it out in a few years.
even you send the PR,It will take years to be merged.

**Comment by hramos** [comment]
@gaodeng we've closed 1,057 pull requests in the last six months, with 730 getting merged into the repository. As I said, we'd be happy to see a PR.

**Comment by Jyrno42** [comment]
I might look into this over the weekend (unless someone beats me to the punch).

**Comment by MartinCura** [comment]
@Jyrno42 I'm not familiar with the RN code but i looked at it for a couple of minutes (i have a OP5T with system font set to OnePlus Slate). (Disclaimer: i'm using RN 0.56 as it's what i have at hand.) The following may help you (or whoever):

- It is reproducible with a Text component with fontWeight set to whatever
- The cut-off seems to be follow the same rules that it would for wrapping a line, but without noticing it has done that
- Cut-off text isn't visible but space for it is still there, so there's some mixed logic: it wraps but still leaves horizontal space as if it hadn't

A bunch of examples and their result:
```javascript
    <View>
      <Button title="I have enough space so nothing's cut-off" style={{ borderRadius: 10 }} />
      <View style={{ position: 'absolute', top: 60, left: 30 }}>
        <Button title="This text will get cut-off!" style={{ position: 'absolute', borderRadius: 10 }} />
      </View>
      <View style={{ top: 80 }}>
        <Text style={{ backgroundColor: 'red', alignSelf: 'center' }}>First second third</Text>
        <Text style={{ backgroundColor: 'red', alignSelf: 'center', fontWeight: '500' }}>
          First second third
        </Text>
        <Text style={{ alignSelf: 'center', fontWeight: '500' }}>0123456789</Text>
        <Text style={{ alignSelf: 'center', fontWeight: '500' }}>01234567890j</Text>
        <Text style={{ backgroundColor: 'red', alignSelf: 'center', fontWeight: '500' }}>
          The following would get cut off longwordfortesting!
        </Text>
        <Text style={{ backgroundColor: 'green', alignSelf: 'center', fontWeight: '500', height: 70 }}>
          The following would get cut off longwordfortesting!
        </Text>
        <Text style={{ backgroundColor: 'red', alignSelf: 'center', fontWeight: '500', height: 70 }}>
          The following would get cut off longerwordmakesmorepace
        </Text>
        <Text style={{ backgroundColor: 'green', alignSelf: 'center' }}>
          The following would get cut off longerwordmakesmorepace
        </Text>
        <Text style={{ alignSelf: 'center' }}>
          Normal behaviour is for text to get wrapped when too long for a single line
        </Text>
      </View>
    </View>
```

![image](https://user-images.githubusercontent.com/14017665/58274493-63264000-7d69-11e9-8825-347e71f8f76d.png)

tl;dr: it seems to me that the font weight (and maybe other things) isn't taken into account to determine correct size for the text box. If i can be of any help just tell me.

**Comment by vnd-ranga** [comment]
> @mykelaballe cc @hramos We've had similar issue(some texts getting cut off in Oppo device). Although it's not a complete solution, we found some hacky-way to solve this problem.
> 
> in short, (at least in our case) just adding empty `space` at the end of the text seems work. my colleague @saystone will explain in detail :)
> 
> for example, use `1.5.3+space` instead of `1.5.3`(space at the end of the string)

Not working in htc 3t

**Comment by akhilsanker** [comment]
Hi @hramos 

Any permanent solution for this bug. Its been almost 2 years since this task is been opened.

Thanks

**Comment by hramos** [comment]
@akhilsanker see my [earlier comment](https://github.com/facebook/react-native/issues/15114#issuecomment-494684348).

**Comment by sauravexodus** [comment]
I fixed it by doing the following in my `index.js`

```
.
.
.
Text.defaultProps = { fontFamily: Platform.select({ android: 'Roboto' }) };
TextInput.defaultProps = { fontFamily: Platform.select({ android: 'Roboto' }) };
.
.
.
AppRegistry.registerComponent(appName, () => App);
```

**Comment by akhilsanker** [comment]
@hramos ,

I wonder, since Instagram is build upon react-native, how this issue been solved in that. 🤔 

Thanks

**Comment by BuriedInTheGround** [comment]
@akhilsanker,

As far as I know Instagram is only partially built upon react-native, and suffers this issue, for example in post's "insights".

**Comment by Jyrno42** [comment]
Anyone know of an avd image that can reproduce this?

**Comment by sauravexodus** [comment]
> Anyone know of an avd image that can reproduce this?

I think you can use Font Installer app on Android and then use `OnePlus Slate` font to reproduce the same. I haven't tried it myself.

**Comment by Jyrno42** [comment]
Wasn't able to repro on the emulator even when I manually replaced all the font files under `/system/fonts` with the files found [here](https://drive.google.com/drive/folders/0B9UzADWnkrLHM0UtZWZTa0poTTA). The replaced font is rendered but the issue does not occur (see image below). I also made sure to match the screen dimensions of OP5T but that did not make any difference. 

![Screenshot_1562067225](https://user-images.githubusercontent.com/235804/60509765-6b618b80-9cd6-11e9-9513-fd300397580b.png)


Can also confirm this is not happening on OnePlus One (A0001) which uses Cynagenmod (so the default font is Roboto).

E.g. I am still looking for a way to repro this either on an emulator, or by finding a person willing to lend me a valid Oneplus device. I do not think I can figure out a way to deal with this blindly.

**Comment by MartinCura** [comment]
> Wasn't able to repro on the emulator even when I manually replaced all the font files under `/system/fonts` with the files found [here](https://drive.google.com/drive/folders/0B9UzADWnkrLHM0UtZWZTa0poTTA). The replaced font is rendered but the issue does not occur (see image below). I also made sure to match the screen dimensions of OP5T but that did not make any difference.
> 
> ![Screenshot_1562067225](https://user-images.githubusercontent.com/235804/60509765-6b618b80-9cd6-11e9-9513-fd300397580b.png)
> 
> Can also confirm this is not happening on OnePlus One (A0001) which uses Cynagenmod (so the default font is Roboto).
> 
> E.g. I am still looking for a way to repro this either on an emulator, or by finding a person willing to lend me a valid Oneplus device. I do not think I can figure out a way to deal with this blindly.

Thanks for trying! I guess it doesn't repro on an emulator.
 I'd try to fix it but i don't have much experience on this. In my case i just switched the font on my OP5T.

**Comment by simondobson** [comment]
I also ran into this bug, I think that it is because the HTC was using the font "OnePlus Slate". If you switch the phone font (settings > display > font) to Roboto then the bug doesn't appear. A way around this was to override the `fontFamily` to be Roboto across the react-native project

**Comment by hammadzz** [comment]
You now have to remove `proptype` from the examples above in later versions of react-native. Here is what works for me in react-native 0.57.8

Enjoy :)

```javascript
// utils.js
// One Plus Fix for Oxygen OS and its painful Slate font truncating on bold text
// https://github.com/facebook/react-native/issues/15114
export const opoFontFix = () => {
  if(Platform.OS !== 'android') {
    return
  }

  const oldRender = Text.render
  Text.render = function (...args) {
    const origin = oldRender.call(this, ...args);
    return React.cloneElement(origin, {
      style: [{fontFamily: 'Roboto'}, origin.props.style]
    })
  }
}
```


```javascript
// App.js
import {opoFontFix} from './utils'

// One Plus Fix for Oxygen OS
opoFontFix()

//... rest of your code
export default class App extends Component<Props> {
}
```

**Comment by urtheone** [comment]
Environment： react-native ：0.59.9
In my project, we used the offline bundle, and i add the command "--reset-cache" in the react-native bundle command ，we fixed it。

**Comment by hammadzz** [comment]
> offline bundle

Tested a beta release on 0.59.10 on One Plus device and it did not work.

**Comment by hammadzz** [comment]
> Environment： react-native ：0.59.9
> In my project, we used the offline bundle, and i add the command "--reset-cache" in the react-native bundle command ，we fixed it。

Tested a beta release on 0.59.10 on One Plus device and it did not work.

**Comment by Jyrno42** [comment]
Found this - https://docs.kobiton.com/devices-list/ looking into their free trial over the weekend. :)

**Comment by urtheone** [comment]
> You now have to remove `proptype` from the examples above in later versions of react-native. Here is what works for me in react-native 0.57.8
> 
> Enjoy :)
> 
> ```js
> // utils.js
> // One Plus Fix for Oxygen OS and its painful Slate font truncating on bold text
> // https://github.com/facebook/react-native/issues/15114
> export const opoFontFix = () => {
>   if(Platform.OS !== 'android') {
>     return
>   }
> 
>   const oldRender = Text.render
>   Text.render = function (...args) {
>     const origin = oldRender.call(this, ...args);
>     return React.cloneElement(origin, {
>       style: [{fontFamily: 'Roboto'}, origin.props.style]
>     })
>   }
> }
> ```
> 
> ```js
> // App.js
> import {opoFontFix} from './utils'
> 
> // One Plus Fix for Oxygen OS
> opoFontFix()
> 
> //... rest of your code
> export default class App extends Component<Props> {
> }
> ```

do you know why is "robot", can anything else replace it ?
can replace 'robot' with '' ?

**Comment by Jyrno42** [comment]
No luck with the remote devices (tried all Oneplus variants from kobiton.com and testin.net).

**Comment by hammadzz** [comment]
@urtheone sure pick whatever font you want to be defaulted in your app if none is specified. Roboto was picked as it has been Android's default system font. Not sure how true that is today.

**Comment by urtheone** [comment]
> @urtheone sure pick whatever font you want to be defaulted in your app if none is specified. Roboto was picked as it has been Android's default system font. Not sure how true that is today.

well, thanks.

**Comment by rsp8055** [comment]
I got this issue in OnePlus 7. After some workaround got to the following solution:

imp: this issue only got for the text which was having { fontWeight: 'bold' } property
 
initially, my Text component was:

<Text style={{
fontSize: 20,
fontWeight: 'bold'
}}>
SOMETHING
</Text> 

After got issue tried solution for all Text component which was having a bold property I download and add (Roboto_Bold) font because by default there is different font in OnePlus 7(OnePlus Slate font) :

<Text style={{
fontSize: 20,
fontFamily: 'Roboto_Bold',
}}>
SOMETHING
</Text>

Hope this will help.

**Comment by dulmandakh** [comment]
This is known issue, and its duplicate of https://github.com/facebook/react-native/issues/25481. Workaround is to set fontFamily.

**Comment by treyp** [comment]
@dulmandakh -- this issue has been around much longer and has much more history. Shouldn't the newer one be closed? Also, setting fontFamily isn't a workaround but actually produces the bug. Avoiding font weights is one workaround, but not a very useful one.

**Comment by gaodeng** [comment]
@dulmandakh  please reopen this issue.

**Comment by dulmandakh** [comment]
I think it doesn't matter how we track the issue, but I find https://github.com/facebook/react-native/issues/25481 has clean technical information about the issue.

**Comment by gaodeng** [comment]
@dulmandakh 
This issue is subscribed by many people. This problem has not been solved. You simply closed it, which will cause the previously subscribed users  unable to track the latest status of this issue.

**Comment by sapjax** [comment]
> @dulmandakh
> This issue is subscribed by many people. This problem has not been solved. You simply closed it, which will cause the previously subscribed users unable to track the latest status of this issue.

They just can't face the fact that they haven't solved this problem for years 🤣.

**Comment by zhenbushidashen** [comment]
I tried add the property 'lineHeight' and it works perfectly on my Oneplus phone

**Comment by MayoudP** [comment]
> I tried add the property 'lineHeight' and it works perfectly on my Oneplus phone

You mean, adding `lineHeight={1}` for example just fixed it ?

**Comment by johnxie** [comment]
Is this happening on all Oneplus phones? Deciding which devices from Oneplus to purchase to test this bug for [Taskade](https://play.google.com/store/apps/details?id=com.taskade.mobile&showAllReviews=true).

**Comment by HemantRMali** [comment]
Adding Space at end of text works for me.

Example: {empIDText + " "} 

before it was like: {empIDText}

**Comment by RichardPK** [comment]
Guys while adding a space at the end of text may 'solve' the issue - this is a totally inappropriate solution for any centred text, which will now be offset.

**Comment by punjasin** [comment]
u need to change the default font off the app maybe add font family will fix the problem the problem is came from the font of the phone itself

**Comment by cmacdonnacha** [comment]
@RichardPK wouldn't forcing the font family of the app to something like "Roboto" be a permanent / good solution?

**Comment by RichardPK** [comment]
@cmacdonnacha yep - @hkxicor 's comment on 5 Nov 2017 https://github.com/facebook/react-native/issues/15114#issuecomment-341988546 was the fix we successfully implemented to fix this in our app.

**Comment by cmacdonnacha** [comment]
@RichardPK thanks. Does that solution change the text globally or will have to do this in each component I use <Text> ?

**Comment by luoyushouchai** [comment]
> You now have to remove `proptype` from the examples above in later versions of react-native. Here is what works for me in react-native 0.57.8
> 
> Enjoy :)
> 
> ```js
> // utils.js
> // One Plus Fix for Oxygen OS and its painful Slate font truncating on bold text
> // https://github.com/facebook/react-native/issues/15114
> export const opoFontFix = () => {
>   if(Platform.OS !== 'android') {
>     return
>   }
> 
>   const oldRender = Text.render
>   Text.render = function (...args) {
>     const origin = oldRender.call(this, ...args);
>     return React.cloneElement(origin, {
>       style: [{fontFamily: 'Roboto'}, origin.props.style]
>     })
>   }
> }
> ```
> 
> ```js
> // App.js
> import {opoFontFix} from './utils'
> 
> // One Plus Fix for Oxygen OS
> opoFontFix()
> 
> //... rest of your code
> export default class App extends Component<Props> {
> }
> ```

Not work in react native 0.59.9.

**Comment by iamadityaaz** [comment]
> this solution works for me, you can try this
> 
> `import React from 'react';`
> `import { `
> ` StyleSheet,` Text,` `} from 'react-native';`
> 
> `const styles = StyleSheet.create({`
> ` defaultFontFamily: {`
> ` fontFamily: 'lucida grande',`
> ` }`
> `});`
> 
> `const oldRender = Text.prototype.render;`
> `Text.prototype.render = function (...args) {`
> ` const origin = oldRender.call(this, ...args);`
> ` return React.cloneElement(origin, {`
> ` style: [origin.props.style, styles.defaultFontFamily]`
> ` });`
> `};`



> @hkxicor
> 
> ```diff
>  import React from 'react';
>  
>  import { StyleSheet, Text } from 'react-native';
>  
>  const styles = StyleSheet.create({
>    defaultFontFamily: {
>      fontFamily: 'lucida grande',
>    },
>  });
>  
>  export default function fixOppoTextCutOff() {
>    const oldRender = Text.prototype.render;
>    Text.prototype.render = function render(...args) {
>      const origin = oldRender.call(this, ...args);
>      return React.cloneElement(origin, {
> -      style: [origin.props.style, styles.defaultFontFamily],
> +      style: [styles.defaultFontFamily, origin.props.style],
>      });
>    };
>  }
> ```

I was getting `render` error
Fixed it by removing `proptype` from the code.

**Comment by Romaiin52** [comment]
I have faced similar issues with several devices (Samsung galaxy s10, Google Pixel..).
The solutions above don't work for me, I fixed it by setting **textBreakStrategy** to **simple** on the Text element.
Hope it can help!

**Comment by lorenc-tomasz** [comment]
The same issue here with OnePlus 5.

**Comment by The0racle** [comment]
> I have faced similar issues with several devices (Samsung galaxy s10, Google Pixel..).
> The solutions above don't work for me, I fixed it by setting **textBreakStrategy** to **simple** on the Text element.
> Hope it can help!

Tried many suggestions, even forked react native as suggested [here](https://github.com/facebook/react-native/issues/24837#issuecomment-514653372), but nothing worked.  Setting **textBreakStrategy** to **simple** did the trick. Thank you!!

**Comment by YesSkyscrapers** [comment]
> I have faced similar issues with several devices (Samsung galaxy s10, Google Pixel..).
> The solutions above don't work for me, I fixed it by setting **textBreakStrategy** to **simple** on the Text element.
> Hope it can help!

THANKS, MAN, GREAT SOLUTION

**Comment by abhisekpadhi** [comment]
Here is one way I created a custom component that standardises `fontFamily`.

Library function:

```
// Custom Text component to standardise font-family
export function StandardText(props: {children: ReactNode}) {
    return <Text style={fontStyles.text}>{props.children}</Text>;
}

const fontStyles = StyleSheet.create({
    text: {
        ...Platform.select({
            android: {fontFamily: 'Roboto'},
        }),
    },
});
```

Implementation: 

```
<StandardText>
          <Text style={styles.nameText}>{this.props.id}</Text>
</StandardText>
```

This solves text cut off issue with OnePlus devices. it should also solve for the devices with non standard fonts.

**Comment by mohit23x** [comment]
i was facing same issue on one plus device, i resolved by creating a custom text component.

```
<View>
<Text style={{width:'100%'}}>text here</Text>
</View>
```

Using only text with 100% width works, but you may face text alignment issues

**Comment by NE-SmallTown** [comment]
For me, you can set `numberOfLines={0}` to fix this, I notice that this will [make](https://github.com/facebook/react-native/blame/f53267f7031dec9f8e61c679bf786d025d634eb6/ReactAndroid/src/main/java/com/facebook/react/views/text/ReactTextView.java#L473) react-native(Android) use `Interger.MAXVALUE` as the maxLine(height) for `Text` since 4 years ago  though I don't know why this could fix the bug

**Comment by MoOx** [comment]
Opened bug that seems to be the one we should voice on https://github.com/facebook/react-native/issues/25481

**Comment by dsteinbach** [comment]
In my case `<Text>` was a child of `flex-direction: row;` and the Text was overflowing outside the parent container and not partially not visible. Adding `<Text style={{ flex: 1 }}> `fix the issue similar to @mohit23x's solution.

**Comment by sankaSanjeeva** [comment]
Seems the problem is affected by the font family. I have set font-family globally. But I had to set again font family in some places to fix this issue.

`fontFamily: 'Rubik-Regular'`

**Comment by abhay27chauhan** [comment]
Hi all, this issue fixes if you add fontWeight explicilty in Text Style

**Comment by justjavac** [comment]
5 years have passed, and this issue still hasn't been resolved. Recently, I developed a new app using React Native, and surprisingly, I still employed the same solution as before https://github.com/facebook/react-native/issues/15114#issuecomment-360029841. 😂

**Comment by fabriziobertoglio1987** [comment]
Hello,

Thanks for the issue. I have been contributing to facebook/react-native for 4 years and specialize in the Text/TextInput components. I currently have 58 facebook/react-native PRs:

https://github.com/facebook/react-native/pulls?q=is%3Apr+author%3Afabriziobertoglio1987+ 

It is the suggested approach from the react-native core team (see this [discussion](https://github.com/react-native-community/discussions-and-proposals/discussions/731#discussioncomment-7651145)).

I'm publishing several fixes for Text and TextInput component in a separate library [react-native-improved](https://github.com/fabriziobertoglio1987/react-native-improved). 

- The Component is based on ReactTextView, it is the same implementation from react-native.
- It will include several bug fixes. 
- You can use the library until the PR is merged and released with react-native. 
- You can use side by side both components from react-native and react-native-improved, depending on where you need the fix.
- I will also publish the PR to facebook/react-native (daily fixes and releases).

The advantages would be:
- Increased engagement with the community, issues are quickly fixed.
- No limitations when merging PRs (react-native main branch is used on facebook marketplace).
- Only Meta employees can merge and review facebook/react-native PRs.
- Allows us to further experiment and develop react-native.

What do you think about this? Should I keep working on this? Thanks

**Comment by fabriziobertoglio1987** [comment]
Maybe is related to https://github.com/facebook/react-native/pull/41770

**Comment by fabOnReact** [comment]
I believe this issue is fixed with https://github.com/facebook/react-native/pull/41770

**Comment by NaOMi-HaNaKi** [comment]
> 你好，
> 
> 感谢您提出问题。我已经为 facebook/react-native 做出了 4 年的贡献，并且专注于 Text/TextInput 组件。我目前有 58 个 facebook/react-native PR：
> 
> https://github.com/facebook/react-native/pulls?q=is%3Apr+author%3Afabriziobertoglio1987+
> 
> 这是react-native核心团队建议的方法（请参阅此[讨论](https://github.com/react-native-community/discussions-and-proposals/discussions/731#discussioncomment-7651145)）。
> 
> 我在单独的库中发布了 Text 和 TextInput 组件的多个修复程序[react-native-improved](https://github.com/fabriziobertoglio1987/react-native-improved)。
> 
> * 该组件基于ReactTextView，它与react-native的实现相同。
> * 它将包括几个错误修复。
> * 您可以使用该库，直到PR与react-native合并并发布。
> * 您可以同时使用react-native 和react-native-improved 中的两个组件，具体取决于您需要修复的位置。
> * 我还将把 PR 发布到 facebook/react-native（每日修复和发布）。
> 
> 优点是：
> 
> * 增加与社区的互动，问题很快得到解决。
> * 合并 PR 时没有限制（facebook 市场上使用的是 React-native 主分支）。
> * 只有 Meta 员工可以合并和审查 facebook/react-native PR。
> * 使我们能够进一步实验和开发react-native。
> 
> 你怎么看待这件事？我应该继续努力吗？谢谢



> [我相信这个问题已通过#41770](https://github.com/facebook/react-native/pull/41770)解决[](https://github.com/facebook/react-native/pull/41770)
Hello，thank you for the solution.
But when I use 'npm install react-native-improved',The console tells me it cannot install and reported an error


> [我相信这个问题已通过#41770](https://github.com/facebook/react-native/pull/41770)解决[](https://github.com/facebook/react-native/pull/41770)



> Hello,
> 
> Thanks for the issue. I have been contributing to facebook/react-native for 4 years and specialize in the Text/TextInput components. I currently have 58 facebook/react-native PRs:
> 
> https://github.com/facebook/react-native/pulls?q=is%3Apr+author%3Afabriziobertoglio1987+
> 
> It is the suggested approach from the react-native core team (see this [discussion](https://github.com/react-native-community/discussions-and-proposals/discussions/731#discussioncomment-7651145)).
> 
> I'm publishing several fixes for Text and TextInput component in a separate library [react-native-improved](https://github.com/fabriziobertoglio1987/react-native-improved).
> 
> * The Component is based on ReactTextView, it is the same implementation from react-native.
> * It will include several bug fixes.
> * You can use the library until the PR is merged and released with react-native.
> * You can use side by side both components from react-native and react-native-improved, depending on where you need the fix.
> * I will also publish the PR to facebook/react-native (daily fixes and releases).
> 
> The advantages would be:
> 
> * Increased engagement with the community, issues are quickly fixed.
> * No limitations when merging PRs (react-native main branch is used on facebook marketplace).
> * Only Meta employees can merge and review facebook/react-native PRs.
> * Allows us to further experiment and develop react-native.
> 
> What do you think about this? Should I keep working on this? Thanks

Hello，thank you for the solution.
But when I use 'npm install react-native-improved',The console tells me it cannot install and reported an error
Looking forward to your recovery

React-Native version 0.70
npm 6
node 18.12.1

Looking forward to your recovery

**Comment by NaOMi-HaNaKi** [comment]
> I believe this issue is fixed with #41770

Hello，thank you for the solution.
But when I use 'npm install react-native-improved',The console tells me it cannot install and reported an error
Looking forward to your recovery

React-Native version 0.70
npm 6
node 18.12.1

Looking forward to your recovery

**Comment by NaOMi-HaNaKi** [comment]
> Hello,
> 
> Thanks for the issue. I have been contributing to facebook/react-native for 4 years and specialize in the Text/TextInput components. I currently have 58 facebook/react-native PRs:
> 
> https://github.com/facebook/react-native/pulls?q=is%3Apr+author%3Afabriziobertoglio1987+
> 
> It is the suggested approach from the react-native core team (see this [discussion](https://github.com/react-native-community/discussions-and-proposals/discussions/731#discussioncomment-7651145)).
> 
> I'm publishing several fixes for Text and TextInput component in a separate library [react-native-improved](https://github.com/fabriziobertoglio1987/react-native-improved).
> 
> * The Component is based on ReactTextView, it is the same implementation from react-native.
> * It will include several bug fixes.
> * You can use the library until the PR is merged and released with react-native.
> * You can use side by side both components from react-native and react-native-improved, depending on where you need the fix.
> * I will also publish the PR to facebook/react-native (daily fixes and releases).
> 
> The advantages would be:
> 
> * Increased engagement with the community, issues are quickly fixed.
> * No limitations when merging PRs (react-native main branch is used on facebook marketplace).
> * Only Meta employees can merge and review facebook/react-native PRs.
> * Allows us to further experiment and develop react-native.
> 
> What do you think about this? Should I keep working on this? Thanks

I found the problem because of "packageManager": "yarn@3.6.1" in the dependency package. Is the yarn version necessary? thanks

**Comment by fabOnReact** [comment]
@NaOMi-HaNaKi sorry. I only support rn 0.73 for now, in the future I may support more versions (feature request [#18](https://github.com/fabOnReact/react-native-improved/issues/18)). I tested with npm on rn 0.73 ([screenshot](https://github.com/facebook/react-native/assets/24992535/c6d5caf9-81d2-4281-944a-c31b99111c1a)) and there are no issues. 

I believe it is the same issue with https://github.com/facebook/react-native/issues/39722 (PR #41770 may fix both). 

The root issue was that the Text width was calculated wrong with StaticLayout from the ReactTextShadowNode measure function. The issue reproduces when the text width is higher of it's parent and needs to be cut with an ellipsis. StaticLayout uses line break strategy to decide where to cut the text. It works well with more than one line, but with one line may remove an entire word and the text width is not the full width of the parent.

Using BoringLayout in that case fixed the issue. Your text is boring, does not include gifs, so it should fix this issue too.

I'll update the issue in 1 week with more info, Thanks a lot.

**Comment by NaOMi-HaNaKi** [comment]
> @NaOMi-HaNaKi sorry. I only support rn 0.73 for now, in the future I may support more versions (feature request [#18](https://github.com/fabOnReact/react-native-improved/issues/18)). I tested with npm on rn 0.73 ([screenshot](https://github.com/facebook/react-native/assets/24992535/c6d5caf9-81d2-4281-944a-c31b99111c1a)) and there are no issues.
> 
> I believe it is the same issue with #39722 (PR #41770 may fix both).
> 
> The root issue was that the Text width was calculated wrong with StaticLayout from the ReactTextShadowNode measure function. The issue reproduces when the text width is higher of it's parent and needs to be cut with an ellipsis. StaticLayout uses line break strategy to decide where to cut the text. It works well with more than one line, but with one line may remove an entire word and the text width is not the full width of the parent.
> 
> Using BoringLayout in that case fixed the issue. Your text is boring, does not include gifs, so it should fix this issue too.
> 
> I'll update the issue in 1 week with more info, Thanks a lot.

Thank you so much!

**Comment by fabOnReact** [comment]
@NaOMi-HaNaKi  Can you reproduce it in react-native 0.73? I tested similar issue with custom font OnePlus Slate, but it does not reproduce (see my comment https://github.com/facebook/react-native/issues/25481#issuecomment-1920835446). Thanks

**Comment by NaOMi-HaNaKi** [comment]
> @NaOMi-HaNaKi Can you reproduce it in react-native 0.73? I tested similar issue with custom font OnePlus Slate, but it does not reproduce (see my comment [#25481 (comment)](https://github.com/facebook/react-native/issues/25481#issuecomment-1920835446)). Thanks

Regarding the rn version, the project specified 0.70, and I could not actively change it to 0.73. It is very strange that the problem I encountered did not cause problems when it was displayed locally, but after it was distributed to the user and opened, the display was incomplete.

In addition, I don’t have a OnePlus device. I encountered a similar problem when using another device, Redmi K30 - Xiaomi's default font. When I adjusted the device font to Big Mac Bold, the content was not fully displayed.And I have tried testing the same font on other Xiaomi phones and the problem did not occur.I started to wonder if the problem was related to rn…

**Comment by NaOMi-HaNaKi** [comment]
A similar situation also appeared in RedMi K30 MIUI13 setting MiSans font, Big Mac thick. All numbers involved will be truncated at the end.

**Comment by ClaudiaSkyeB** [comment]
Any progress on whether there will be a bugfix for this?

**Comment by JPStrydom** [comment]
This seems to happen whenever the text is a multiple of 5 characters (5, 10, 15, etc.). Our solution to this, which feels very dirty, is:
```js
const fixedText = useMemo(
  () => (text && !(text?.length % 5) ? ` ${text} ` : text),
  [text]
);
```
Really strange that this issue has been around for so long.

**Comment by lchenfox** [comment]
I fixed it to add nested outer `Text` wrapper like:

```
import React, {memo} from "react";
import {Text} from "react-native";
import PropTypes from "prop-types";

function CLText(props) {
	const {children, allowFontScaling = false} = props;
	// Note that `allowFontScaling` must be set between two `Text` below.
	return (
		<Text allowFontScaling={allowFontScaling}>
			<Text {...props} allowFontScaling={allowFontScaling}>
				{children}
			</Text>
		</Text>
	);
}

CLText.propTypes = {
	style: PropTypes.oneOfType([
		PropTypes.object,
		PropTypes.array,
	]),
	allowFontScaling: PropTypes.bool,
	lineBreakMode: PropTypes.string,
	numberOfLines: PropTypes.number,
	onPress: PropTypes.func,
	textBreakStrategy: PropTypes.string,
};

export default memo(CLText);
```

**Comment by johnsimeroth** [comment]
Hi, I'm just another voice calling into the abyss for help. We have the same issue on both pixel and galaxy devices. Hopefully https://github.com/facebook/react-native/pull/41770 fixes this and gets merged soon... 🤞

**Comment by umrashrf** [comment]
Dang, a MeetRyde.com customer just reported that on Samsung S22 Ultra some texts are truncated. That's odd because I tested the app on Galaxy S9+ which is also Android and it works perfectly fine.

**Comment by majih93** [comment]
@umrashrf That is probably due to the new default font Samsung applied on its devices after OneUI 6 (The new font name is One UI Sans). Galaxy S9+ doesnt have OneUI 6 so you would not have seen this issue on your device.

**Comment by umrashrf** [comment]
@majih93 that's what I guessed reading the internet and someone suggested to hardcode the font with font-family property which I did and will test soon at Samsung Experience Center :)

**Comment by AspireOne** [comment]
Most of the previous solutions alone did not work for me, but after combining them, this worked for me in latest expo 50, React Native 0.73.6

1. Patching the global Text component using [react-fast-hoc](https://github.com/XantreDev/react-fast-hoc/) to add properties:
	- style: [{ fontFamily: "Roboto" }, props.style],
	- textBreakStrategy: "simple"
	- numberOfLines: 0

2. Overriding the Text.render function to:
	- add a Roboto font

`patch-text-component.ts`
```ts
import * as RN from "react-native";
import { createTransformProps } from "react-fast-hoc";

export default () => {
  if (RN.Platform.OS !== "android") {
    return;
  }

  const styles = RN.StyleSheet.create({
    font: { fontFamily: "Roboto" },
  });

  const transform = createTransformProps(
    (props) => ({
      ...props,
      style: [styles.font, props.style],
      textBreakStrategy: "simple",
      numberOfLines: 0,
    }),
    {
      namePrefix: "Reset.",
      mimicToNewComponent: false,
    },
  );

  Object.assign(RN.Text, transform(RN.Text));
};
```

`patch-text-render.ts`
```ts
import { Platform, Text, StyleSheet } from "react-native";
import React from "react";

export default () => {
  if (Platform.OS !== "android") {
    return;
  }

  // @ts-expect-error
  const oldRender = Text.render ?? Text.prototype?.render;
  const style = StyleSheet.create({ font: { fontFamily: "Roboto" } });
  if (!oldRender) {
    console.error(
      "Text.render or Text.prototype.render is not defined, cannot patch font.",
    );
    return;
  }

  if (Text.prototype?.render) {
    Text.prototype.render = function (...args: any[]) {
      const origin = oldRender.call(this, ...args);
      return React.cloneElement(origin, {
        style: [style.font, origin.props.style],
      });
    };
    // @ts-expect-error
  } else if (Text.render) {
    // @ts-expect-error
    Text.render = function (...args: any[]) {
      const origin = oldRender.call(this, ...args);
      return React.cloneElement(origin, {
        style: [style.font, origin.props.style],
      });
    };
  } else {
    console.error(
      "Text.render or Text.prototype.render is not defined, cannot patch font.",
    );
  }
};
```

Usage:
`src/app/_layout.tsx`
```ts
// ...
try {
  patchTextComponent();
  patchTextRender();
} catch (e) {
  console.error("Failed to patch text component", e);
}
// ...
```

**Comment by mieradi** [comment]
Seeing this issue on a Pixel 8 device as well. Has there been any progress?

**Comment by Dsalvat596** [comment]
Seeing this issue on Android devices (e.g, Pixel 4) , when toggling  `Bold Text` to true in accessibility settings on the device.  The last word/character/letter of some  <Text> Components are cut off.  Adding a space ` ` to the  string seems to fix the UI (as that space is getting cut off instead).

But there must be a better solution.

**Comment by Faizey** [comment]
Any updates on this? I'm seeing this issue on Samsung S23

**Comment by rohankm** [comment]
Im also facing this on some devices... how to fix it?

**Comment by bennett-davis** [comment]
I saw this fix somewhere in a different thread somewhere I think, but it worked for me:

```
const androidStyles = useMemo(() => {
    // Can change this to isAndroid && boldTextEnabled
    // if AccessibilityInfo ever supports it
    if (Platform.OS === 'android') {
      return {
        borderRightWidth: 2 * fontScale,
        borderColor: colors.transparent,
      };
    }
    return {};
  }, [fontScale]);
```

The code comment is because in my app it only breaks if the user has a different default font selected and has bold text enabled. The border width multiplying by font scale makes it work with different device font sizes.

**Comment by jlo1** [comment]
Started experiencing this issue with a custom font on various Android devices (including Pixel & SGalaxy), **only after** updating from RN 0.76 --> 0.79.4 

Note: bold was not enabled on these devices as in some cases mentioned above

Wonder if @NickGerleman 's recent fixes could be relevant here?
* https://github.com/facebook/react-native/pull/51183
* https://github.com/facebook/react-native/pull/51206

**Comment by huanguolin** [comment]
Our app started experiencing this issue after upgrading from version 0.72 to 0.77 and enabling the new architecture. We've already received multiple user reports.
They all encountered the problem after adjusting the system font size or enabling bold text.
This issue did not occur in the previous version.

**Comment by alexandrermo** [comment]
We have this issue using expo SDK 53 with react-native 0.79.5. We have view this bug on emulator API 36, Android 16 "Baklava".

**Comment by roadmanfong** [comment]
After upgrading React Native from 0.73 to 0.76, 
some Android users experienced this issue. 
Setting allowFontScaling to false on Text component or specifying fontFamily (see [related discussion](https://github.com/facebook/react-native/issues/15114#issuecomment-2422537975)) can mitigate the problem, but these are not perfect solutions.

We would appreciate any help or suggestions regarding this issue.

**Comment by mykel-unwritten** [comment]
Crazy to think that i posted this issue 8 years ago and until now the issue still occurs

**Comment by wuwb** [comment]
have this issue on react-native 0.76.9

**Comment by KestasVenslauskas** [comment]
same on `0.80.x` , `0.81.x`, `0.82.x` versions...

**Comment by Kimchoky** [comment]
same 0.79.5.
galaxy S20 with android 13 -> Normal
galaxy S25 Ultra with android 16 -> strings inside the button cut off

**Comment by Mukhunyeledzi-Muthuphei** [comment]
Having the same issue, looks fine on S20 FE but on Note 20 text is cut off throughout the app
